components/public-api-server/pkg/apiv1/workspace_test.go

```diff
@@ -184,6 +188,122 @@ func TestWorkspaceService_GetOwnerToken(t *testing.T) {
+			Input: protocol.WorkspaceInfo{
+				Workspace: &protocol.Workspace{
+					BaseImageNameResolved: "foo:bar",
+					ID:                    "gitpodio-gitpod-isq6xj458lj",
+					OwnerID:               "fake-owner-id",
+					ContextURL:            "https://github.com/gitpod-io/gitpod",
+					Context: &protocol.WorkspaceContext{
+						NormalizedContextURL: "https://github.com/gitpod-io/gitpod.git",
+						Title:                "tes ttitle",
+					},
+					Description: "test description",
+				},
+				LatestInstance: &protocol.WorkspaceInstance{
+					ID:           "f2effcfd-3ddb-4187-b584-256e88a42442",
+					IdeURL:       "https://gitpodio-gitpod-isq6xj458lj.ws-eu53.gitpod.io/",
+					CreationTime: "2022-07-12T10:04:49+0000",
+					WorkspaceID:  "gitpodio-gitpod-isq6xj458lj",
+					Status: &protocol.WorkspaceInstanceStatus{
+						Conditions: &protocol.WorkspaceInstanceConditions{
+							Failed:            "nope",
+							FirstUserActivity: "2022-07-12T10:04:49+0000",
+							Timeout:           "nada",
+						},
+						Message: "has no message",
+						Phase:   "running",
+						Version: 42,
+					},
+				},
+			},
+			Expectation: Expectation{
+				Result: &v1.ListWorkspacesResponse_WorkspaceAndInstance{
+					Result: &v1.Workspace{
+						WorkspaceId: "gitpodio-gitpod-isq6xj458lj",
+						OwnerId:     "fake-owner-id",
+						Context: &v1.WorkspaceContext{
+							ContextUrl: "https://github.com/gitpod-io/gitpod",
+							Details: &v1.WorkspaceContext_Git_{
+								Git: &v1.WorkspaceContext_Git{
+									NormalizedContextUrl: "https://github.com/gitpod-io/gitpod",
+								},
+							},
+						},
+						Description: "test description",
+					},
+					LastActiveInstances: &v1.WorkspaceInstance{
+						InstanceId:  "f2effcfd-3ddb-4187-b584-256e88a42442",
+						WorkspaceId: "gitpodio-gitpod-isq6xj458lj",
+						CreatedAt:   timestamppb.New(must(time.Parse(time.RFC3339, "2022-07-12T10:04:49Z"))),
+						Status: &v1.WorkspaceInstanceStatus{
+							StatusVersion: 42,
+							Phase:         v1.WorkspaceInstanceStatus_PHASE_RUNNING,
+							Conditions: &v1.WorkspaceInstanceStatus_Conditions{
+								Failed:            "nope",
+								Timeout:           "nada",
+								FirstUserActivity: timestamppb.New(must(time.Parse(time.RFC3339, "2022-07-12T10:04:49Z"))),
+							},
+							Message:   "has no message",
+							Url:       "https://gitpodio-gitpod-isq6xj458lj.ws-eu53.gitpod.io/",
+							Admission: v1.AdmissionLevel_ADMISSION_LEVEL_OWNER_ONLY,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.Name, func(t *testing.T) {
+			var (
+				act Expectation
+				err error
+			)
+			act.Result, err = convertWorkspaceInfo(&test.Input)
+			if err != nil {
+				act.Error = err.Error()
+			}
+
+			if diff := cmp.Diff(test.Expectation, act, protocmp.Transform()); diff != "" {
+				t.Errorf("unexpected convertWorkspaceInfo (-want +got):\n%s", diff)
+			}
+		})
+	}
+}
+
+func FuzzConvertWorkspaceInfo(f *testing.F) {
```

Comment:
nice!

Reply:
haven't found anything with this yet though :)

Reply:
Just a really good implementation :)